cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.cpp

```diff
@@ -0,0 +1,421 @@
+    return sycl::event{};
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduction_rm_cw<Float, BinaryOp, UnaryOp>::operator()(const Float* input,
+                                                                  Float* output,
+                                                                  const std::int64_t width,
+                                                                  const std::int64_t height,
+                                                                  const std::int64_t stride,
+                                                                  const BinaryOp& binary,
+                                                                  const UnaryOp& unary,
+                                                                  const event_vector& deps) const {
+    const auto method = propose_method(width, height);
+    return this->operator()(method, input, output, width, height, stride, binary, unary, deps);
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduction_rm_cw<Float, BinaryOp, UnaryOp>::operator()(
+    const typename reduction_rm_cw<Float, BinaryOp, UnaryOp>::reduction_method method,
+    const Float* input,
+    Float* output,
+    const std::int64_t width,
+    const std::int64_t height,
+    const BinaryOp& binary,
+    const UnaryOp& unary,
+    const event_vector& deps) const {
+    return this->operator()(method, input, output, width, height, width, binary, unary, deps);
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduction_rm_cw<Float, BinaryOp, UnaryOp>::operator()(const Float* input,
+                                                                  Float* output,
+                                                                  const std::int64_t width,
+                                                                  const std::int64_t height,
+                                                                  const BinaryOp& binary,
+                                                                  const UnaryOp& unary,
+                                                                  const event_vector& deps) const {
+    const auto method = propose_method(width, height);
+    return this->operator()(method, input, output, width, height, width, binary, unary, deps);
+}
+
+#define INSTANTIATE(F, B, U) template class reduction_rm_cw<F, B, U>;
+
+#define INSTANTIATE_FLOAT(B, U)                \
+    INSTANTIATE(double, B<double>, U<double>); \
+    INSTANTIATE(float, B<float>, U<float>);
+
+INSTANTIATE_FLOAT(min, identity)
+INSTANTIATE_FLOAT(min, abs)
+INSTANTIATE_FLOAT(min, square)
+
+INSTANTIATE_FLOAT(max, identity)
+INSTANTIATE_FLOAT(max, abs)
+INSTANTIATE_FLOAT(max, square)
+
+INSTANTIATE_FLOAT(sum, identity)
+INSTANTIATE_FLOAT(sum, abs)
+INSTANTIATE_FLOAT(sum, square)
+
+#undef INSTANTIATE_FLOAT
+
+#undef INSTANTIATE
+
+#endif
+
+} // namespace oneapi::dal::backend::primitives
```

Comment:
Long file with different kernel implementations. Is it reasonable to divide it on many `cpp`s?

Reply:
fixed